Update to tree-sitter 0.23

hendrikvanantwerpen opened the pull request:
Tree-sitter released version 0.23 recently. This contains a breaking, but ultimately very good change to how parser bindings work (https://github.com/tree-sitter/tree-sitter/pull/3069). The TLDR of that change is that parsers do not depend on tree-sitter anymore, but instead on a shard (and supposedly very stable) tree-sitter-language crate. As a result, clients are free to chose their tree-sitter version as the parser ABI is supported. Library and parser versions are less tighly coupled, and it should no longer be necessary to move all the parsers to the next tree-sitter version in lock-step to be able to upgrade the library.

Most of the changes in this PR are from running tree-sitter generate. I'll add comments on other changes I did to explain why I thought they are necessary, or where I'm not sure I did it correctly.

hendrikvanantwerpen commented on bindings/rust/lib.rs:
```diff
@@ -1,52 +1,53 @@
-// pub const LOCALS_QUERY: &'static str = include_str!("../../queries/locals.scm");
-pub const TAGS_QUERY: &'static str = include_str!("../../queries/tags.scm");
+pub const HIGHLIGHTS_QUERY: &str = include_str!("../../queries/highlights.scm");
+pub const INJECTIONS_QUERY: &str = include_str!("../../queries/injections.scm");
```
I uncommented this because the injections file existed already in the repo.

hendrikvanantwerpen commented on package.json:
```diff
@@ -19,7 +24,8 @@
   },
   "scripts": {
-    "test": "tree-sitter test",
+    "build": "npx tree-sitter-cli generate --no-bindings",
```
I changed this to make the scripts less reliant on the system state. Running `npx tree-sitter-cli` instead of `tree-sitter` will ensure the version specified in the `devDependencies` is used.

hendrikvanantwerpen reviewed (commented):
Some additional comments on the changes:

- I manually bumped the version. Let me know if that's not the right way, though.
- I did not update the CHANGELOG yet. Let me know if that should be part of this PR or happens as part of the release process.

Let me know if this works for you, or I need to change anything!

jonatanklosko commented on src/parser.c:
We can remove the generated src/ files from the diff, we relay on CI to automatically generate and push these :)

jonatanklosko reviewed (approved):
@hendrikvanantwerpen thank you for the PR!

@the-mikedavis if you think it looks good, I can release 0.3.0 after merging :)

hendrikvanantwerpen commented:
Let me quickly run the binding tests locally to ensure I didn't mess those up. I'll report back within the hour on that.

hendrikvanantwerpen commented:
Thanks for the review @jonatanklosko! I've removed the generated srcs as per your suggestion, and fixed some issues I found with the bindings. This should be good now.
